### ISSUE TITLE
Silent Hill - Shattered Memories - Memory Stick Not Inserted issue

### ISSUE THREAD

**Issue by alkaris2** [hrydgard/ppsspp#13781]
### What happens?
While playing Silent Hill - Shattered Memories I cannot save the game from the in-game Save Menu, because it's telling me that it can't find the Save Data that I created when I started a new game. It's right there in directory where it first created it, but for some reason the emulator cannot see the data exists for the game to read and save to.

![ULUS10450_00000](https://user-images.githubusercontent.com/73155407/102448913-0a24a800-4083-11eb-8612-dd25647b14e1.png)

I can only use Save States to save the game, not the in-game Save option. I tried checking the System settings for Memory Stick Inserted option to check that it is in fact Enabled from the menu. Toggling on and off doesn't seem to change whether if it detects the Save Data that already exists.

I've tried other stuff like creating appropriate directories for SAVEDATA directories like so;

```
~/.config/ppsspp/PSP/memstick/PSP/SAVEDATA/<GAME_ID_DIR>

~/.config/ppsspp/memstick/PSP/SAVEDATA/<GAME_ID_DIR>
```
doesn't read or detect any of these directories.

### What *should* happen?

You should be able to save game data from the in-game Save option as normal by reading and saving that data from SAVEDATA in `~/.config/ppsspp/PSP/SAVEDATA` directory. 

### What hardware, operating system, and PPSSPP version?  On desktop, GPU matters for graphical issues.

**CPU:** Intel™ Core® i7-7700K
**RAM:** Trident Z RGB 32Gb
**GPU:** Asus GTX 1070

**OS:** Manjaro Linux - 5.9.11-3-MANJARO
**PPSSPP Version:** v1.10.3

**Comment by LunaMoo** [comment]
Probably just a duplicate of #7244

Try restarting emulation and loading the game from in-game save(not using savestates), then try to save. If it does save, then it's the mentioned issues. Games have protection/validation for memstick and abusing savestates can break it by simply loading a savestate made before last in-game savedata was made. This is an user time travel issue which can't really be fixed, but games can usually be easily patched via cheats to disable that memstick protection allowing to save anyway.

**Comment by alkaris2** [comment]
@LunaMoo I've tried this, I've tried restarting emulation and it does not work. I created a new test file to test it out, but if you try to load the initial save data that it creates, it says the save data it created is corrupt, when the save file isn't corrupt, the save file is sane.

The game is just simply broken with no way to save the game the properly way from in-game. Because new file shouldn't be read as corrupt if the game created it and put it into the SAVEDATA directory as it would on real PSP hardware.  No Save States, just save file of the game itself. -- I don't know of any  game codes or cheats for SH: Shattered Memories that could bypass this memstick protection if this game has that.  So for the time being the only way of saving the game is through Save States unless there's a patch code for the game or something the emulator can provide for the game to recognise the save data correctly.

**Comment by unknownbrackets** [comment]
A few suggestions:
 * Try the latest git build.
 * Make sure you're not using a hacked version of the ISO.  Some versions were patched to work better on old custom firmware, but this makes them run worse on PPSSPP.  You want a clean rip of your disc.

That said, this isn't the first time this has been reported:
https://forums.ppsspp.org/showthread.php?tid=23072&pid=103356#pid103356

The weird thing is, plenty of people have played the game, reported issues with other things, and not had saving issues.  This makes me think it's related to a patched ISO.

-[Unknown]

**Comment by MiguelCardonaG** [comment]
The same problem also happens on Android, if you make any changes within the Settings section, when saving this problem comes out:

![Screenshot_20201227-184839_PPSSPP](https://user-images.githubusercontent.com/65356692/103301563-f6256280-4a01-11eb-8843-3aae899308cb.jpg)

I have tried it with a clean ISO and with three different smartphones and in all gives the same problem.

**Comment by unknownbrackets** [comment]
It happens when you change settings?  That means saving works as long as you don't?

What if you go to settings, change nothing, and return to the game?

-[Unknown]

**Comment by MiguelCardonaG** [comment]
@unknownbrackets If I don't change anything within Settings, and I give Cancel there's no error.

**Comment by unknownbrackets** [comment]
Interesting.  I wonder if it could be checking the mtime of all files on the memory stick, really?

-[Unknown]

**Comment by MiguelCardonaG** [comment]
> Interesting. I wonder if it could be checking the mtime of all files on the memory stick, really?
> 
> -[Unknown]

The funny or strange thing is that if you want to save the game later, that error occurs.

**Comment by Saramagrean** [comment]
Try with v1.16.6 is work fine on PC but still happen on Android.

https://github.com/hrydgard/ppsspp/assets/26355472/8ee9ef60-049d-4cbc-90fd-6039d72b2911

**Comment by sum2012** [comment]
Confirm v1.16.6 is work fine on PC but still happen on Android.
How to test  MemstickFixedFree  option in android ?
I cannot find compat.ini in android.

**Comment by hrydgard** [comment]
It's built-in the APK on Android, but you can add an extra compat.ini in memstick/PSP/SYSTEM for quick tests.

**Comment by sum2012** [comment]
unlucky, MemstickFixedFree not work
PPSSPP v1.16.6-235-g3611b4d94 Windows 64 bit log: (Create a new save, Change option,then save again)
https://gist.github.com/sum2012/7ea46cca9f4c3f3079f37d707165c462

**Comment by Marcost9** [comment]
Right now, I'm playing: Silent Hill Sharrind Memory on the app. Over the years, I've noticed some changes in the scenery that don't exist in the other version of the console, namely: shadows, detail in the scenery itself. I understood at that moment that there was a limitation of the console, but it was not the console but the application. But I didn't give up, I, being a bit picky, took the courage to fix the problem and it worked. So, checking the option: "software rendering" but another problem appeared, which was a drop in FPS/speed. Perhaps with future updates this graphical problem will appear less. The application has portacia.eu, I intend to study a little programming to solve these problems

**Comment by blurrred** [comment]
Same problem on Manjaro Linux.
Just gonna try the PCSX2 version and see if it works better.

**Comment by sum2012** [comment]
Maybe relate to #19275

**Comment by hrydgard** [comment]
Looking at sceUtility output, After LIST, GETSIZE, FILES, WRITEDATASECURE and then LIST again, it gives up on Android while it happily continues on Windows. Must be some filesystem weirdness.. Not sure why it wouldn't recognize its own files in LIST... 

Think I'll have to add a bunch of logging to see what's different.

**Comment by hrydgard** [comment]
Ok, so this fails on macOS as well, and it doesn't look like it's failing writing. 

I'm starting to suspect some more memory issues with the savegame encryption (fixed one bug a while ago, there might be more).

**Comment by hrydgard** [comment]
Got some time to valgrind it, and it indeed warns about a use of an uninitialized value in the encryption code, which could definitely cause this type of issue..

**Comment by sum2012** [comment]
ppsspp v1.17-1-893-gg5c26aa4c96 android version
The same problem also happens on Android, if you make any changes within the Settings section, when saving this problem comes out

**Comment by hrydgard** [comment]
Yes, this is still unsolved (and quite mysterious). Updates will be posted here if it gets fixed, unless a commit is related to save data or similar, no need to respond with "still not working".

**Comment by IrfanH495** [comment]
sorry if outside the issue, I deactivate BlockTransferAllowCreateFB
flashlight turned off
![Screenshot_2024-08-18-13-27-33-861](https://github.com/user-attachments/assets/5d0d5d38-a135-41e1-9978-816fcecfd081)
flashlight turned on
![Screenshot_2024-08-18-13-27-42-144](https://github.com/user-attachments/assets/9d1fc56f-1f7b-4fc7-b4e6-cd44bdc2d504)
flashlight turned on while opening the door
![Screenshot_2024-08-18-13-28-35-184](https://github.com/user-attachments/assets/03aff134-7386-43e3-b541-4f6c5981764c)
flashlight turned on and open the phone
![Screenshot_2024-08-18-13-32-09-244](https://github.com/user-attachments/assets/5e8311f0-bdac-49ad-bd67-ec60b1327a36)
flashlight turned on and open camera
![Screenshot_2024-08-18-13-32-43-232](https://github.com/user-attachments/assets/d2219e1e-a17c-4e7e-8a0a-8e1df277cc58)
this is the first time i see readbacks 3

**Comment by hrydgard** [comment]
I've been messing around some more with this, adding more logging, but the API-surface behavior of the game seems to be pretty much identical between Windows (where it works) and Linux (where it doesn't).. no easily detectable errors. No idea what's going on here.

I first thought it had to be some kind of file system semantic, like case sensitivity, but the filenames contain only numbers, and it has problems both on old-school Android shared storage and on scoped storage. I really don't get it :/

**Comment by Panderner** [comment]
Can reducing the Memory Stick size can help this issue?

**Comment by DavidAriyan** [comment]
This is a serious bug. You can try play the ps2 version with AetherSX2. It's very playable. Same as Origins. The only difference the psp version has is lower textures and smaller size. Ps2 version is 2 GB and psp version can be compressed to around 700mb. You decide.

**Comment by sum2012** [comment]
@hrydgard Where do the build Variant in android studio ?
edit: forget upload image.

<img width="927" height="462" alt="Image" src="https://github.com/user-attachments/assets/d3c9137f-8bae-42f0-a94c-67753ba3fcf5" />

**Comment by hrydgard** [comment]
@sum2012 See the three dots to the left? Click it, Build Variants should be there. It's kinda silly that it doesn't show up by default as it's very important.

**Comment by hrydgard** [comment]
~~Oh, I think sync needs to succeed first. Click Try Again on the blue banner.~~ (actually that would be a catch 22... hm)

Hm, otherwise try View -> Tool Windows -> Build Variants in the menu

**Comment by sum2012** [comment]
still no.

<img width="932" height="463" alt="Image" src="https://github.com/user-attachments/assets/ab770bb6-7965-4f00-88bc-cfab11de838a" />

**Comment by sum2012** [comment]
@hrydgard I updated Android Studio , Used Gemini Agent help me to build Android debug build. 
edit: Success

**Comment by hrydgard** [comment]
great!

**Comment by sum2012** [comment]
Update error message  "Please re-insert the Memory Stick containing your save profile data".

**Comment by sum2012** [comment]
I think the problem is that something wrong in producing save file on Android.

**Comment by sum2012** [comment]
Important data by minimax AI, 
1.
Silent Hill uses crypto mode=5 (not modes 14-17 as originally assumed)
2.
Dialog operations use modes 15 (READDATASECURE) and 17 (WRITEDATASECURE)

**Comment by sum2012** [comment]
Find by minimax ai from logcat
java.lang.IllegalArgumentException: Failed to determine if primary:Download/iso/PSP/SAVEDATA/ULUS10450STORE.000 is child of primary:Download/iso: 
java.io.FileNotFoundException: Missing file for primary:Download/iso/PSP/SAVEDATA/ULUS10450STORE.000

edit:The problem is that Android_CreateDirectory expects a tree URI, but the code is passing a document URI
Now I can see the root cause of the permission issue! The problem is that while the app correctly grants permission for the selected directory, it's still using the default storage path instead of the user-selected directory.
SecurityException: Permission Denial: writing com.android.externalstorage.ExternalStorageProvider uri content://com.android.externalstorage.documents/tree/primary%3ADownload/document/primary%3ADownload from pid=12059, uid=11407 requires android.permission.MANAGE_DOCUMENTS, or grantUriPermission()

edit2: wrong change for the AI

**Comment by senpaikazu2018-collab** [comment]
![Image](https://github.com/user-attachments/assets/7f9aa6f1-22f2-4232-ac6f-a397e729b3c0)

https://github.com/user-attachments/assets/42a87689-c50d-4432-97b4-cf4bf6f716ac

The error still occur such a great game but I'm hard stuck on save dialog

**Comment by TomieAi** [comment]
![Image](https://github.com/user-attachments/assets/a16cb692-c6c0-47db-ad1a-70087e8d3f79)

**Comment by antipirgos** [comment]
Assassins Creed Bloodlines and Monster Hunter 3rd had saving problem too but it's fixed with  cheat code hack.

**Comment by fcrwr** [comment]
Seems to check creation time, doesn't like that ctime of entries returned by `SavedataParam::GetList` changes, on linux `stat` function returns time of last change of file's metadata, not creation time. `statx` function can get creation time, but does it work on android? Either way a simple hack seems to fix the issue, but requires the game to be started with a new profile.
```diff
diff --git a/Common/File/DirListing.cpp b/Common/File/DirListing.cpp
index 8ada1dffa0..8cb6b61997 100644
--- a/Common/File/DirListing.cpp
+++ b/Common/File/DirListing.cpp
@@ -128,7 +128,7 @@ bool GetFileInfo(const Path &path, FileInfo * fileInfo) {
 	fileInfo->exists = true;
 	fileInfo->atime = file_info.st_atime;
 	fileInfo->mtime = file_info.st_mtime;
-	fileInfo->ctime = file_info.st_ctime;
+	fileInfo->ctime = 0;
 	fileInfo->access = file_info.st_mode & 0x1ff;
 	// HACK: approximation
 	if (file_info.st_mode & 0200)
```
Or ctime can be zeroed in `SavedataParam::GetList`

**Comment by sum2012** [comment]
@fcrwr don't work for overwrite save in android

**Comment by fcrwr** [comment]
@sum2012 try this from new game started with new profile, without using previous save states, works on my android device:
```diff
diff --git a/Common/File/AndroidStorage.cpp b/Common/File/AndroidStorage.cpp
index 1163d50426..de7076fe85 100644
--- a/Common/File/AndroidStorage.cpp
+++ b/Common/File/AndroidStorage.cpp
@@ -220,7 +220,7 @@ static bool ParseFileInfo(std::string_view line, File::FileInfo *fileInfo) {
 
 	// We don't have better information, so let's just spam lastModified into all the date/time fields.
 	fileInfo->mtime = lastModified;
-	fileInfo->ctime = lastModified;
+	fileInfo->ctime = 0;
 	fileInfo->atime = lastModified;
 	return true;
 }
diff --git a/Common/File/DirListing.cpp b/Common/File/DirListing.cpp
index 8ada1dffa0..8cb6b61997 100644
--- a/Common/File/DirListing.cpp
+++ b/Common/File/DirListing.cpp
@@ -128,7 +128,7 @@ bool GetFileInfo(const Path &path, FileInfo * fileInfo) {
 	fileInfo->exists = true;
 	fileInfo->atime = file_info.st_atime;
 	fileInfo->mtime = file_info.st_mtime;
-	fileInfo->ctime = file_info.st_ctime;
+	fileInfo->ctime = 0;
 	fileInfo->access = file_info.st_mode & 0x1ff;
 	// HACK: approximation
 	if (file_info.st_mode & 0200)
```

**Comment by hrydgard** [comment]
@fcrwr that is a really interesting find! That indicates that the game is doing some seemingly-misguided check on the creation time, but one that happens to work on hardware.

I'm not exactly sure what we should do here, 0 seems kinda risky for the sake of other games, I'll test a few things tomorrow, after first confirming the fix.

**Comment by sum2012** [comment]
Confirm works on my android device
> [@sum2012](https://github.com/sum2012) try this from new game started with new profile, without using previous save states, works on my android device:
> 
> diff --git a/Common/File/AndroidStorage.cpp b/Common/File/AndroidStorage.cpp
> index 1163d50426..de7076fe85 100644
> --- a/Common/File/AndroidStorage.cpp
> +++ b/Common/File/AndroidStorage.cpp
> @@ -220,7 +220,7 @@ static bool ParseFileInfo(std::string_view line, File::FileInfo *fileInfo) {
>  
>  	// We don't have better information, so let's just spam lastModified into all the date/time fields.
>  	fileInfo->mtime = lastModified;
> -	fileInfo->ctime = lastModified;
> +	fileInfo->ctime = 0;
>  	fileInfo->atime = lastModified;
>  	return true;
>  }
> diff --git a/Common/File/DirListing.cpp b/Common/File/DirListing.cpp
> index 8ada1dffa0..8cb6b61997 100644
> --- a/Common/File/DirListing.cpp
> +++ b/Common/File/DirListing.cpp
> @@ -128,7 +128,7 @@ bool GetFileInfo(const Path &path, FileInfo * fileInfo) {
>  	fileInfo->exists = true;
>  	fileInfo->atime = file_info.st_atime;
>  	fileInfo->mtime = file_info.st_mtime;
> -	fileInfo->ctime = file_info.st_ctime;
> +	fileInfo->ctime = 0;
>  	fileInfo->access = file_info.st_mode & 0x1ff;
>  	// HACK: approximation
>  	if (file_info.st_mode & 0200)

**Comment by hrydgard** [comment]
Experimenting here, I can reproduce the problem on Windows by setting ctime=atime=mtime. 

I can also "fix" it by forcing ctime to 0 as above. However I tried setting ctime to `mtime - 1000`, and that didn't help. I also tried small offsets, also didn't help.

I think maybe it compares ctime before and after it's saving, checking that it stayed the same... Which would be why it doesn't work to return various offsets of mtime as ctime, and why overriding it to 0 every time does work.

Unfortunately on Android with scoped storage we cannot get a real ctime, unless we store ctimes manually in some database...

So I think we need a hack indeed. The question is if this will affect older saves, I'll have to test that... if it does, I don't really know a good fix..

**Comment by hrydgard** [comment]
It does appear to work fine with existing savegames.

I'm going to implement a compat flag.

**Comment by TomieAi** [comment]
> It does appear to work fine with existing savegames.
> 
> I'm going to implement a compat flag.

In android do we need some extra steps to enable that?

**Comment by hrydgard** [comment]
No, just use the latest development builds or wait for 1.20 (soon)

**Comment by antipirgos** [comment]
I tested all versions of the game using ppsspp_v1.19.3-1855-ga32f550871.apk only the Japanese version won't save game, USA/EU work without problems. The only problem is when flashlight is turned on, game work very slow, which is a big problem in Silent Hill Shattered Memories and Silent Hill Origins.

**Comment by hrydgard** [comment]
@antipirgos can you tell me the game ID of the japanese version? is it ULJM-05564? (long press the game icon in PPSSPP to see it)

**Comment by antipirgos** [comment]
![Image](https://github.com/user-attachments/assets/c296f118-04bb-4159-8c66-399076c91c61)

**Comment by hrydgard** [comment]
Ok, will enable for that ID.

Note: No mentions of where to download ISOs please. We have rules against that for good reasons.

**Comment by antipirgos** [comment]
Sorry about that. Is ULJM-05564 version work in new update???

**Comment by hrydgard** [comment]
It should work in the next devbuild.
